src/ckb/channel.rs

```diff
@@ -2518,43 +2550,99 @@ impl ChannelActorState {
                 )));
             }
         };
-        let tx_builder = TransactionBuilder::default()
-            .cell_deps(get_cell_deps_by_contracts(vec![Contract::CommitmentLock]))
-            .input(
-                CellInput::new_builder()
-                    .previous_output(self.get_funding_transaction_outpoint())
-                    .build(),
-            );
 
-        // TODO: Check UDT type, if it is ckb then convert it into u64.
-        let local_value = (self.to_local_amount - local_shutdown_fee) as u64;
-        let remote_value = (self.to_remote_amount - remote_shutdown_fee) as u64;
-        debug!(
-            "Building shutdown transaction with values: {} {}",
-            local_value, remote_value
-        );
-        let local_output = CellOutput::new_builder()
-            .capacity(local_value.pack())
-            .lock(local_shutdown_script)
-            .build();
-        let remote_output = CellOutput::new_builder()
-            .capacity(remote_value.pack())
-            .lock(remote_shutdown_script)
-            .build();
-        let outputs = self.order_things_for_musig2(local_output, remote_output);
-        let tx_builder = tx_builder.set_outputs(outputs.to_vec());
-        let tx_builder = tx_builder.set_outputs_data(vec![Default::default(), Default::default()]);
-        let tx = tx_builder.build();
-        let message = get_funding_cell_message_to_sign(
-            u64::MAX,
-            self.get_funding_transaction_outpoint(),
-            &tx,
-        );
-        debug!(
-            "Building message to sign for shutdown transaction {:?}",
-            hex::encode(message.as_slice())
+        let mut contracts = vec![Contract::CommitmentLock];
+        if self.funding_udt_type_script.is_some() {
+            // TODO(yukang): we need to find corresponding dep cells for UDT
+            contracts.push(Contract::SimpleUDT);
+        }
+        let cell_deps = get_cell_deps_by_contracts(contracts);
+        let tx_builder = TransactionBuilder::default().cell_deps(cell_deps).input(
+            CellInput::new_builder()
+                .previous_output(self.get_funding_transaction_outpoint())
+                .build(),
         );
-        Ok((tx, message))
+
+        if let Some(type_script) = &self.funding_udt_type_script {
+            debug!(
+                "UDT local_amount: {}, remote_amount: {}",
+                self.to_local_amount, self.to_remote_amount
+            );
+            let local_output_data = self.to_local_amount.to_le_bytes().pack();
+            let dummy_output = CellOutput::new_builder()
+                .lock(local_shutdown_script.clone())
+                .type_(Some(type_script.clone()).pack())
+                .build();
+            let required_capacity = dummy_output
```

Comment:
When openning a UDT channel, the local CKB amount must be enough to hold the commitment tx outputs and shutdown tx outputs. So the local output capacity must be the local CKB amount - fee but not the minimum required capacity.

Reply:
resolved it with https://github.com/contrun/ckb-pcn-node/pull/37/commits/7249c9f41d9d5d3c1bf6cd24cb83a6cf42b2d091